### PR TITLE
disable   test_client_server and double_fit tests for concrete knn

### PR DESCRIPTION
**Pull request by kcelia**
Issue explained: https://github.com/zama-ai/concrete-ml-internal/issues/4014

In this PR:

- I changed some variable names, I found them confusing.
- Disabled the test_client_server and double_fit tests for concrete knn
- I took the opportunity to remove an ugly loop in base.py and replace it with slicing. I thought that slicing wasn't allowed in FHE.

refs #https://github.com/zama-ai/concrete-ml-internal/issues/4014
closes #https://github.com/zama-ai/concrete-ml-internal/issues/4009
closes #https://github.com/zama-ai/concrete-ml-internal/issues/4017
Closes https://github.com/zama-ai/concrete-ml-internal/issues/4001

### CODE REVIEW

File: tests/deployment/test_client_server.py [hunk 1]
```diff
@@ -106,23 +103,18 @@ def test_client_server_sklearn(
...
-    fhe_circuit = model.compile(
-        x_train, default_configuration, **extra_params, show_mlir=(n_bits <= 8)
-    )
-
-    if get_model_name(model) == "KNeighborsClassifier":
-        # Fit the model
-        with warnings.catch_warnings():
-            # Sometimes, we miss convergence, which is not a problem for our test
-            warnings.simplefilter("ignore", category=ConvergenceWarning)
-            model.fit(x, y)
```
Comment (kcelia) [inline tests/deployment/test_client_server.py:118]: It seems that this part is the source of the problem !
To be confirmed.

File: tests/deployment/test_client_server.py [hunk 2]
```diff
@@ -255,28 +247,26 @@ def client_server_simulation(x_train, x_test, model, default_configuration):
 
     # Back to the client
 
-    # Decrypt and de-quantize the result
+    # Decrypt, de-quantize and post-processed the result
     y_pred_on_client_quantized = fhemodel_client.deserialize_decrypt(serialized_result)
     y_pred_on_client_dequantized = fhemodel_client.deserialize_decrypt_dequantize(serialized_result)
```
Comment (kcelia) [inline tests/deployment/test_client_server.py:252]: deserialize_decrypt_dequantize has postprocessing as well.
maybe we should rename it, it's confusing

File: tests/deployment/test_client_server.py [hunk 3]
```diff
@@ -255,28 +247,26 @@ def client_server_simulation(x_train, x_test, model, default_configuration):
...
 
     # Get the y_pred_model_server_clear
 
     # Predict based on the model we are testing
     qtest = model.quantize_input(x_test)
-    y_pred_model_server_ds_quantized = model.fhe_circuit.encrypt_run_decrypt(qtest)
```
Comment (kcelia) [inline tests/deployment/test_client_server.py:266]: y_pred_model_server_ds_quantized are neither the outputs of the client nor the outputs of the server. But the outputs of the dev model, right ?

**Review by andrei-stoian-zama — APPROVED** [review]

File: src/concrete/ml/sklearn/base.py [hunk 4]
```diff
@@ -1975,11 +1975,9 @@ def scatter1d(x, v, indices):
                     r = p
 
             # Return only the topk indexes
-            topk_labels = []
-            for i in range((self.n_neighbors)):
-                topk_labels.append(labels[i])
+            topk_labels = fhe_array(labels[:k])
 
-            return fhe_array(topk_labels)
+            return topk_labels
```
Comment (kcelia) [inline src/concrete/ml/sklearn/base.py:1980]: Slicing is allowed in FHE.

So I just replaced the loop by fhe_array(labels[:k])

File: tests/sklearn/test_sklearn_models.py [hunk 5]
```diff
@@ -225,6 +225,10 @@ def check_correctness_with_sklearn(
 def check_double_fit(model_class, n_bits, x_1, x_2, y_1, y_2):
     """Check double fit."""
 
+    if get_model_name(model_class) == "KNeighborsClassifier":
```
Comment (kcelia) [inline tests/sklearn/test_sklearn_models.py:228]: Same error as the  "tests/deployment/test_client_server.py".

File: tests/deployment/test_client_server.py [hunk 6]
```diff
@@ -85,6 +82,10 @@ def test_client_server_sklearn(
 ):
     """Tests the encrypt decrypt api."""
 
+    # if get_model_name(model_class) == "KNeighborsClassifier":
```
Comment (RomanBredehoft) [inline tests/deployment/test_client_server.py:85]: is this supposed to be uncommented ?

Comment (RomanBredehoft) [reply tests/deployment/test_client_server.py:85]: if so please add `FIXME:` before the link and most of all put "https://github.com/zama-ai/concrete-ml-internal/issues/4014" instead of #4019 !

File: tests/deployment/test_client_server.py [hunk 7]
```diff
@@ -106,23 +107,18 @@ def test_client_server_sklearn(
...
     max_bit_width = fhe_circuit.graph.maximum_integer_bit_width()
     print(f"Max width {max_bit_width}")
 
     # Check that the FHE execution is correct.
     # With a global_p_error of 1/100_000 we only allow one run.
+
+    # Check the real FHE prediction vs the clear ones
+    # The simulated predictions are not considered in this test
     check_is_good_execution_for_cml_vs_circuit(x_test, model, simulate=False, n_allowed_runs=1)
+    # Check client/server FHE predictions vs the FHE predictions of the dev model
```
Comment (RomanBredehoft) [inline tests/deployment/test_client_server.py:121]: could you add add a space above this comment for better readability ?

File: tests/deployment/test_client_server.py [hunk 8]
```diff
@@ -106,23 +107,18 @@ def test_client_server_sklearn(
...
-            # Sometimes, we miss convergence, which is not a problem for our test
-            warnings.simplefilter("ignore", category=ConvergenceWarning)
-            model.fit(x, y)
+    fhe_circuit = model.compile(x_train, default_configuration, **extra_params, show_mlir=False)
 
     max_bit_width = fhe_circuit.graph.maximum_integer_bit_width()
     print(f"Max width {max_bit_width}")
 
     # Check that the FHE execution is correct.
     # With a global_p_error of 1/100_000 we only allow one run.
```
Comment (RomanBredehoft) [inline tests/deployment/test_client_server.py:116]: should these two comments be removed then ?

File: tests/deployment/test_client_server.py (hunk 2 above)
Comment (RomanBredehoft) [reply tests/deployment/test_client_server.py:256]: yes I agree, maybe this could be done along https://github.com/zama-ai/concrete-ml-internal/issues/3045. In any case, I think you could open an issue for this API change proposal and add it to https://github.com/zama-ai/concrete-ml-internal/issues/3995

File: tests/deployment/test_client_server.py (hunk 3 above)
Comment (RomanBredehoft) [reply tests/deployment/test_client_server.py:266]: output of the dev circuit yes

File: tests/sklearn/test_sklearn_models.py [hunk 9]
```diff
@@ -225,6 +225,13 @@ def check_correctness_with_sklearn(
...
+        # https://github.com/zama-ai/concrete-ml-internal/issues/4019
```
Comment (RomanBredehoft) [inline tests/sklearn/test_sklearn_models.py:229]: same, #4019 is closed + could you add `FIXME:` before the link please !

File: tests/deployment/test_client_server.py (hunk 7 above)
Comment (kcelia) [reply tests/deployment/test_client_server.py:121]: bad rebase !
Thanks

File: tests/deployment/test_client_server.py (hunk 8 above)
Comment (kcelia) [reply tests/deployment/test_client_server.py:116]: Are you talking about:
# Check that the FHE execution is correct.
# With a global_p_error of 1/100_000 we only allow one run.

Comment (RomanBredehoft) [reply tests/deployment/test_client_server.py:116]: yes I am, since you added comments below, or maybe merge them together ?

File: tests/sklearn/test_sklearn_models.py [hunk 10]
```diff
@@ -225,6 +225,13 @@ def check_correctness_with_sklearn(
 def check_double_fit(model_class, n_bits, x_1, x_2, y_1, y_2):
     """Check double fit."""
 
+    if get_model_name(model_class) == "KNeighborsClassifier":
+        # https://github.com/zama-ai/concrete-ml-internal/issues/4014
```
Comment (RomanBredehoft) [inline tests/sklearn/test_sklearn_models.py:229]: please add `FIXME:` before any issue links

File: tests/deployment/test_client_server.py [hunk 11]
```diff
@@ -85,6 +82,10 @@ def test_client_server_sklearn(
 ):
     """Tests the encrypt decrypt api."""
 
+    if get_model_name(model_class) == "KNeighborsClassifier":
+        # https://github.com/zama-ai/concrete-ml-internal/issues/4014
```
Comment (RomanBredehoft) [inline tests/deployment/test_client_server.py:86]: please add `FIXME:` before any issue links + a  small comment explaining what needs to be done / fixed

File: tests/deployment/test_client_server.py (hunk 7 above)
Comment (RomanBredehoft) [reply tests/deployment/test_client_server.py:121]: (I think you forgot to add it)

**Review by RomanBredehoft — CHANGES_REQUESTED** [review]
still have a few comments ! Also, since you close the flaky issues, could you at least add the reproducible commands with seeds found in them in https://github.com/zama-ai/concrete-ml-internal/issues/4014 so that we can better debug next time ?

File: tests/deployment/test_client_server.py (hunk 7 above)
Comment (kcelia) [reply tests/deployment/test_client_server.py:121]: > bad rebase !
> Thanks

My bad, it was meant for the above comment lol not for this one.
Yes I'll add a space.

File: tests/deployment/test_client_server.py (hunk 6 above)
Comment (kcelia) [reply tests/deployment/test_client_server.py]: Yes it wasn't supposed to be commented.

I replaced by 4014, thanks @RomanBredehoft

File: tests/deployment/test_client_server.py (hunk 11 above)
Comment (kcelia) [reply tests/deployment/test_client_server.py:86]: alright

File: tests/deployment/test_client_server.py (hunk 8 above)
Comment (kcelia) [reply tests/deployment/test_client_server.py:116]: I'll merge them

File: tests/deployment/test_client_server.py (hunk 2 above)
Comment (kcelia) [reply tests/deployment/test_client_server.py:256]: I added 3045 to the API change proposal list

**Review by RomanBredehoft — APPROVED** [review]
thanks !
